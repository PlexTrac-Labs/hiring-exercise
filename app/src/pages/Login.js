import React, { useRef, useState, useEffect } from "react";
import axios from "axios";
import CredentialWarning from "../components/CredentialWarning";

const LOCAL_STORAGE_KEY = "app.user";

export default function Login() {
  const BASE_URL = "http://localhost:5000"; //todo figure out how to have this just reference path

  //Need a state to store userId
  const [userId, setUserId] = useState([]);
  const [isWrongLogin, setIsWrongLogin] = useState(false);

  //References to Password and Name in form
  const nameRef = useRef();
  const passwordRef = useRef();

  //Add userId to cookie once known, so we can use it in rest of site.
  useEffect(() => {
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(userId));
  }, [userId]);

  function handleLoginSubmit(e) {
    e.preventDefault(); // Prevent html from sending the request itself
    const name = nameRef.current.value;
    const password = passwordRef.current.value;
    if (name === "" || password === "") return;

    //Lets try this out, see how clean it is
    axios
      .post(`${BASE_URL}/authenticate`, {
        username: name,
        password: password
      })
      .then(response => {
        setUserId(response.data);
        window.location = "/profile";
        console.log(response.data);
      })
      .catch(error => {
        setIsWrongLogin(true);
      });
  }

  return (
    <div>
      <form className="login" onClick={handleLoginSubmit}>
        <label>
          Name:
          <input ref={nameRef} type="text" name="name" required />
        </label>
        <br />
        <label>
          Password:
          <input ref={passwordRef} type="text" name="password" required />
        </label>
        <br />
        <input type="submit" />
        {isWrongLogin ? <CredentialWarning /> : null}
      </form>
    </div>
  );
}
