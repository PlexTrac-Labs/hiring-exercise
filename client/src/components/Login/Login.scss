.login-wrapper {
  color: white;
  text-align: center;

  .login-form {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 16px;

    > * {
      margin-top: 8px;
    }

    .username-input {
      color: white;
      width: 250px;
      text-align: start;
      input {
        color: white;
      }
    }
    .password-input {
      color: white;
      width: 250px;
      input {
        color: white;
      }
    }
    .login-btn {
      color: white;
    }
  }
}
