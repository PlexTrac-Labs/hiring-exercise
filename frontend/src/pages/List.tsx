import { Typography } from "@mui/material";
import React from "react";

export const List = () => {
  return <Typography>List</Typography>;
};
