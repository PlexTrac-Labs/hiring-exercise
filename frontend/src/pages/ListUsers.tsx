import { Grid, List, Typography } from "@mui/material";
import React, { useEffect, useState } from "react";
import { CreateUser } from "../components/CreateUser";
import { EditDialog } from "../components/EditDialog";
import { UserListItem } from "../components/UserListItem";
import { User } from "../util/user/types";
import { deleteUser, getAllUsers } from "../util/user/user";

export const ListUsers = () => {
  const [openEditDialog, setOpenEditDialog] = useState(false);
  const [users, setUsers] = useState<User[]>([]);
  const [selectedUser, setSelectedUser] = useState<User>({
    _id: "",
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: "",
    birthYear: 0,
    favColor: "",
    admin: false,
  });

  const handleDelete = (id: string, index: number) => {
    deleteUser(id);
    users.splice(index, 1);
    setUsers([...users]);
  };

  const handleUserClick = (user: User) => {
    setSelectedUser(user);
    setOpenEditDialog(true);
  };

  useEffect(() => {
    let getAll = async () => {
      let response = await getAllUsers();
      setUsers(response);
    };
    getAll();
  }, []);

  return (
    <Grid
      data-testid="listUsers"
      container
      direction="column"
      rowSpacing={2}
      alignItems="center"
      justifyContent="center"
    >
      <Grid item xs={12}>
        <Typography variant="h3">Users</Typography>
        <EditDialog
          user={selectedUser}
          open={openEditDialog}
          setUsers={setUsers}
          onClose={() => setOpenEditDialog(false)}
        />
      </Grid>
      <Grid item xs={12}>
        <List>
          <CreateUser setUsers={setUsers} />
          {users.map((user, index) => (
            <UserListItem
              key={user._id}
              user={user}
              handleClick={handleUserClick}
              handleDelete={() => handleDelete(user._id, index)}
            />
          ))}
        </List>
      </Grid>
    </Grid>
  );
};
