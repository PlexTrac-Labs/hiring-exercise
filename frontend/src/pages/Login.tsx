import {
  Alert,
  Box,
  Button,
  FormControl,
  Grid,
  TextField,
  Typography,
} from "@mui/material";
import { styled } from "@mui/system";
import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { AUTH_TOKEN } from "../helpful/constants";
import { authenticate } from "../util/user/user";

const LoginBox = styled(Box)(({ theme }) => ({
  display: "flex",
  flexDirection: "column",
  padding: 15,
  //backgroundColor: "gray",
  justifyContent: "center",
  alignItems: "center",
  //height: "100vh",
}));

export const Login = () => {
  const navigate = useNavigate();
  const [error, setError] = useState(false);

  const handleSubmit = async (event: any) => {
    event.preventDefault();
    const username = event.target.username.value;
    const password = event.target.password.value;
    try {
      const auth = await authenticate(username, password);
      localStorage.setItem(AUTH_TOKEN, auth.auth_token);
      navigate("list");
    } catch (error) {
      setError(true);
    }
  };

  return (
    <LoginBox>
      <form onSubmit={handleSubmit}>
        <Grid
          container
          direction="column"
          rowSpacing={2}
          alignItems="center"
          justifyContent="center"
        >
          <Grid item xs={12}>
            <Typography variant="h3">Login</Typography>
          </Grid>
          {error && <Alert severity="error">Invalid Username/Password</Alert>}

          <Grid item xs={12}>
            <FormControl>
              <TextField
                required
                id="username"
                variant="outlined"
                label="Username"
              />
            </FormControl>
          </Grid>
          <Grid item xs={12}>
            <FormControl>
              <TextField
                required
                id="password"
                variant="outlined"
                label="Password"
                type="password"
              />
            </FormControl>
          </Grid>
          <Grid item xs={12}>
            <Button color="primary" variant="contained" type="submit">
              Login
            </Button>
          </Grid>
        </Grid>
      </form>
    </LoginBox>
  );
};
