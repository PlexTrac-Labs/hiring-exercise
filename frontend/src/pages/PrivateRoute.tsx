import React from "react";
import { Navigate, Outlet } from "react-router-dom";
import { AUTH_TOKEN } from "../helpful/constants";

export const PrivateRoute = () => {
  const loggedIn = !!localStorage.getItem(AUTH_TOKEN);

  return loggedIn ? <Outlet /> : <Navigate to="/" />;
};
