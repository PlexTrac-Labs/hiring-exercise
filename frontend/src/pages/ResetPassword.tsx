import { Grid, Typography } from "@mui/material";
import React from "react";

export const ResetPassword = () => {
  return (
    <Grid
      container
      direction="column"
      rowSpacing={2}
      alignItems="center"
      justifyContent="center"
    >
      <Grid item xs={12}>
        <Typography variant="h3">Reset Password</Typography>
      </Grid>
    </Grid>
  );
};
