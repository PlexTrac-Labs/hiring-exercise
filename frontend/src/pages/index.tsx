import React from "react";
import { List } from "./List";
import { Login } from "./Login";
import { User } from "./User";

export type PageType = {
  title: string;
  route: string;
  comp: React.FunctionComponent;
};

const Pages: PageType[] = [
  {
    title: "Login",
    route: "/",
    comp: Login,
  },
  {
    title: "List",
    route: "/list",
    comp: List,
  },
  {
    title: "User",
    route: "/user",
    comp: User,
  },
];

export default Pages;
