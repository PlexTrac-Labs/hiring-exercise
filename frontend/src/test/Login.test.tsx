import { cleanup, render } from "@testing-library/react";
import React from "react";
import { Login } from "../pages/Login";

const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...(jest.requireActual("react-router-dom") as any),
  useNavigate: () => mockedUsedNavigate,
}));

afterEach(cleanup);

it("renders", async () => {
  const { asFragment } = render(<Login />);
  expect(asFragment()).toMatchSnapshot();
});
