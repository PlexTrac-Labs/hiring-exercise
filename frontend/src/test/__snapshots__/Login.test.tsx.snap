// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined MuiGrid-direction-xs-column css-1yhrxth-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <h3
          class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
        >
          Login
        </h3>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="username"
              id="username-label"
            >
              Username
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="username"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    Username *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="password"
              id="password-label"
            >
              Password
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="password"
                required=""
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    Password *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Login
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
